Add Keccak variable length input spec (#147)
appliedzkp/zkevm-specs: specs/keccak256/variable-length-input.md

# Keccak256 For Variable Length Input

> We describe the circuit that can process input of arbitrary size.

Reviewer: ```suggestion
We describe the circuit that can process inputs of an arbitrary size.
```

## Plain Keccak behavior

> Keccak256 use a simple iterating sponge construction to handle a variable length input. The sponge absorbs 17 words(136 bytes) of inputs, applies the `Keccak-f` permutation, then absorb next 17 words. Once all parts of the inputs are absorbed, the sponge squeezes out the 32 bytes for the output.
> 
> If the last part of the inputs is less than 136 bytes, a padding must be applied. The padding is a multi-rate padding, it pads single bit 1 followed by minimum number of bits 0 followed by a single bit 1, such that the result is 136 bytes.

Reviewer: ```suggestion
Keccak256 uses a simple iterating sponge construction to handle a variable length input. The sponge absorbs 17 words(136 bytes) of inputs, applies the `Keccak-f` permutation, and then absorb next 17 words. Once all parts of the inputs are absorbed, the sponge squeezes out the 32 bytes for the output.

If the last part of the input is less than 136 bytes, a padding must be applied. The padding is a multi-rate padding, it pads single bit 1 followed by minimum number of bits 0 followed by a single bit 1, such that the result is 136 bytes.
```

## Circuit behavior

### Lookup Region

This region splits the input to multiple parts, each part corresponds to their `Keccak-f` permutation round.
This is also a lookup table for the other circuits to lookup the Keccak256 input to the output.

Define:

We use `Round` for the cells of the following columns in a row, it represents the information in a `keccak_f` round.

The `curr: Round` and `next: Round` represent the current and the next row.

Columns:

- `hash_id` Sequential number for grouping different hash
- `input_len` Length for correct padding
- `input` 136 bytes to be absorbed in this round
- `acc_len` How many length we have absorbed
- `acc_input` Accumulatd bytes by random linear combination (in big-endian order)
- `output` The base-2 `state[:4]` output from this round `keccak_f`
- `is_end_result` This flag indicates the hash output is final. The purpose is for external circuit to lookup.

Selector:

- `is_last_round_of_circuit` Last round of the whole circuit (not hash), to avoid `next` wrap around.

> #### Checks

Reviewer: I think we are missing a check for all of the rounds that will come after the padded one and which are essentially 0.

For all of these, we need to constrain the bytes to be 0 right? Or it does not affect since in the big circuit we will "dissable the permutation" when round > final_round_for_padding?

Author: I think I have an idea for 1 now. Instead of complicating the design to handle unused permutations, the prover should add a dummy hash job `hash_id=3` the use it for 0 inputs for 97 permutations.

Reviewer: Yep, my concern was for 1. I'm not sure I follow you. Following @barryWhiteHat 's idea on that part, he purposed to do the following:

- If permutation contains initial input bytes -> `inp_len <= acc_len` for the last byte of the permutation:
1) ConditionallySelect(prev_output_state  || 0's) into actual state for the permutation computation
2) ConditionallySelect(actual_permutation_bytes || 0's) into actual permutation input
3) `final_result[0..N] += !is_last_permutation * out_state[0..N]
4) `running_sum_last_byte == output` NOTE that 3 and 4 are the same but with different formats.

I can elaborate it more. But will be something like saying, you conditionally copy to the next permutation as inputs `(state and next_inputs)` or a bunch of 0's. And you also conditionally add to `final_result` the result of the permutation depending on if it is the last one or not.

Point 2) sounds good as we discussed! :)

> 1. `hash_id` is sequential `next.id - curr.id in [0, 1]`

Reviewer: I'm not sure if this is exactly needed here. This sounds more like a way to check that permutations are from the same hash operation. 

But that's already guaranteed by the accomulated length counters.

Reviewer: We know this already due to the condition `input_len == acc_len` no?

We also know when to change when `acc_len_prev != acc_len_cur`

Author: Not really, `acc_len` is bumped 136 each row. So it wouldn't exactly equal to `input_len`.

Reviewer: Ohh, so `acc_len` is for permutation, not for all of the input bytes? I thought we had a variable that was couting the length for all of the bytes.
If it's not the case, yes, that's a way to do it with the id :)

> 2. If we are not in the last round of the circuit or last round of permutation

Reviewer: ```suggestion
2. If we are not in the last round of the circuit
```

Not sure what the second part refers. But if it refers to the last round of the circuit too, I recommend to remove this as is more confusing.

Reviewer: Also, would be nice to know if this refers to the last round that requires padding (last round with real witness) or simply the last round of the entire algorithm which might be full of 0's as the input might be much shorter and it is just used for padding constraints.

>    1. Checks the `input_len` is  `curr.input_len === next.input_len`

Reviewer: ```suggestion
   1. `curr.input_len === next.input_len`
```

   2. `next.acc_len === curr.acc_len + 136`
   3. `next.acc_len <= curr.input_len`

>    4. `next.acc_input === curr.acc_input * (r**136) + RLC(input, r)`

Reviewer: I think this falls into the error of using RLC together with the witness within the same column. Is it possible? 
ping @han0110

Reviewer: Maybe I'm mistaken as this is not used to perform an RLC later on.

IIRC, `  4. next.acc_input === curr.acc_input * (r**136) + RLC(input, r)` is checking that we're computing the RLC of the entire input correctly. 

if so, I think I was wrong with my concern and all is fine!

3. If we are in the last round of the circuit or in last round of permutation
   1. Checks the accumulation ends here `assert (curr.input_len - curr.acc_len) in range(136)`
   2. Clear the variables:  `next.acc_len === 0`, `next.acc_input === 0`
4. `is_end_result === next.id - curr.id`

#### Lookup

To lookup the Keccak256 input to the output, query the following columns:

- `is_end_result`: bool
- `input`: RLC
- `output`: RLC

> and constrain `is_end_result === 1`

Author: @ed255 FYI this part might be the one you're looking for

### Padding Region

> Note that we define a new `acc_len` which increament byte by byte, where the `acc_len` in the lookup region bumps by 136 bytes.

Reviewer: ```suggestion
Note that we define a new `acc_len` which increments byte by byte, where the `acc_len` in the lookup region bumps by 136 bytes.
```

#### Plain behavior: The padding rule

```python
def get_padding(acc_len) -> bytes:
    """
    output big-endian bytes
    """
    # note that padlen is at maximum 135
    padlen = 136 - acc_len 
    if padlen == 0:
        # no pad
        return []
    elif padlen == 1:
        # pad 0b10000001
        return [0x81]
    elif 1 < padlen < 136:
        return [0x80] + ([0x00] * int(padlen - 2)) + [0x01]
    else:
        raise ValueError("unreachable")
```

#### Circuit

> The Padding Region is a 136-row region.

Reviewer: I started implementing like 17 rows of 8bytes + 1 word for each. So 17 rows of 9 columns (with a few extra columns for the accomulators). Makes it much easier for the constraint appliance.

Can we copy into another region the required data to make the lookups? Or will increase the costs too much?

Reviewer: Ohhh. 

I started coding like:
- 17 rows of 9 columns
- Copy each byte to the padding_config region where we have the following columns:
- `byte`
- `total_len`
- `acc_len`
- `pad_len`
- `running_sum`

And is in this region where we will copy all bytes from offset 0-136 or even 0-MAX_INP_LEN as they have to be sequential for the entire hash.

WDYT?

Reviewer: The constraints are the same. The only thing that changes is how we layout the data. I did it like this as simplifies a lot more the things form the impl side. 

But maybe this is not inline with your proposal, and then I'm happy to refactor

Author: Make sense to me. As long as the padding_config is 1 col 136 rows it should be fine. It's okay to copy `byte` col in padding_config to a 17-rows-9 columns gadget for other processes.

Reviewer: Can't only be this. Needs some padding extra fields as I said.
Otherwise, you miss info to make the constraints between bytes.

Config looks like:
```rust
/// Constrains each byte of the keccak input to be padded correctly.
#[derive(Debug, Clone)]
pub(crate) struct PaddingBuilderConfig<F> {
    q_enable: Selector,
    // Contains:
    // - `byte` at Rotation(0)
    // - `pad_byte` at Rotation(1)
    byte: Column<Advice>,
    total_len: Column<Advice>,
    acc_len: Column<Advice>,
    // Contains:
    // - `pad_len` at Rotation(0)
    // - `pad_len_inv` at Rotation(1)
    pad_len: Column<Advice>,
    running_sum: Column<Advice>,
}
```
I can change this and avoid using rotations, having more columns and less rows.
Feel free to decide on this! :)

Author: Can we do "constraints between bytes" by copying them to 17 word-gadgets?

Reviewer: I'm not sure I follow you here. 
The idea i was following (what I understood in our last call) was something along the lines of:

- Witness all of the bytes and padd correctly in `PermutationConfig`.
- From the previous config, we copy the info required to check that a byte is correctly padded into `PaddingBuilderConfig`.
- Finally, once all these constraints are applied, we copy each 8 bytes to the WordBuilder together with the word they form to constraint word construction correctness as well as that the bytes are indeed between [0-255].

Then we return the RLC accomulated for each byte of the permutation + the resulting State words.

Author: Resolved offline. We agreed to go naive and defer the optimization later.
The optimization is about merging `byte` and `pad_byte` and using `Rotation(2)` to reduce columns and with the cost of increasing rotation, which is more cost-efficient in the aggregation circuit. But merging the columns would make debugging gates harder and optimization at this stage is a bit early.

- `byte` individual byte of the input in big-endian
- `input_len` Length for correct padding
- `acc_len` How many bytes we have processed
- `condition_80_inv` The inverse of `input_len - acc_len` or 0.

> - `padded_byte` Mostly the same as the original `byte` but padded

Reviewer: Not sure what `padded_byte` can be other than itself. You mean that it could be 0?

I say it because padded word would maybe make a bit more sense, but byte does not much to me.

Author: For example, if the input length is 4 and it is `[0xff, 0xff, 0xff, 0xff]`. It would looks like

| byte | padded_byte |
| ---| --- |
|0xff|0xff|
|0xff|0xff|
|0xff|0xff|
|0xff|0xff|
|0x00|0x80|
|...|...|
|0x00|0x01|

Reviewer: So it's the permutation input what is padded. The byte can only be: 
- Itself
- 0
- 0x80
- 0x01
- (Special case) 0x81

Author: Correct. I would consider 0 as the part of "Itself" though.

#### Prover behavior

If the prover uses less `Round`s than the circuit provides, the prover should add a dummy hash that has an input size to use all rest of the `Round`s.

### Padding Region

Note that we define a new `acc_len` which increments byte by byte, where the `acc_len` in the lookup region bumps by 136 bytes.

#### Plain behavior: The padding rule

```python
def get_padding(acc_len) -> bytes:
    """
    output big-endian bytes
    """
    # note that padlen is at maximum 135
    padlen = 136 - acc_len 
    if padlen == 0:
        # no pad
        return []
    elif padlen == 1:
        # pad 0b10000001
        return [0x81]
    elif 1 < padlen < 136:
        return [0x80] + ([0x00] * int(padlen - 2)) + [0x01]
    else:
        raise ValueError("unreachable")
```

#### Circuit

The Padding Region is a 136-row region.

- `byte` individual byte of the input in big-endian
- `input_len` Length for correct padding
- `acc_len` How many bytes we have processed
- `condition_80_inv` The inverse of `input_len - acc_len` or 0.
- `padded_byte` Mostly the same as the original `byte` but padded
- `is_pad_zone` A flag to define the rows that `byte` should be 0

> | offset | byte | input_len | acc_len | condition_80_inv | padded_byte | is_pad_zone |

Reviewer: Padding table is missing accumulator value which is the accomulative RLC.

Author: Do we want to do them in this table or should we copy them to another RLC gadget as proposed in point 6? Because to accumulate RLC here we need to place the randomness in this column and complicate the logic.

Reviewer: We can have the randomness in an instance column as @han0110 suggested. 
The only issue I see is that we have a lot of rows, which will mean that we have A LOT of instance rows too. And generally we want to have as less PI as possible.

Reviewer: We can otherways just copy the randomness once, and copy_constraint it from row to row. So that we make sure that the randomness used is always the same queried from the initial instance column.

Author: Added the RLC column for `byte` in [87f13e6](https://github.com/appliedzkp/zkevm-specs/pull/147/commits/87f13e620cd9c159e2d031bba1d844d88c777920)

2. The padding `0x80`, `0x01`, or `0x81` are placed at the correct place.

We apply two different checks on the 0~134-th rows and the 135th row.

1. For 0-th row
   1. `input_len` is copied from the Lookup Region
   2. `acc_len` is copied from the Lookup Region
   3. `curr.is_pad_zone === 0`
2. For all rows
   1. If `is_pad_zone` then `byte === 0`. `is_pad_zone * byte === 0`
3. For 0~134-th rows
   1. `next.input_len === curr.input_len`
   2. `next.acc_len === curr.acc_len + 1`
   3. Inverse check for `curr.condition_80_inv`
   4. If `curr.input_len - curr.acc_len` is 0, pad `0x80`: `curr.padded_byte === curr.byte + (1 - (curr.input_len - curr.acc_len) * curr.condition_80_inv) * 0x80`
4. For 1~135th rows
   1. Set `is_pad_zone` to 1 if we entered. `curr.is_pad_zone === prev.is_pad_zone + (1 - (curr.input_len - curr.acc_len) * curr.condition_80_inv)`
5. For the 135th row
   1. Same as the previous 0x80 pad, but pad 0x01 if we are in the pad zone. `curr.padded_byte === curr.byte + (1 - (curr.input_len - curr.acc_len) * curr.condition_80_inv) * 0x80 + is_pad_zone * 0x01`
6. Use another RLC gadget to check `byte` can be running summed up to `input` in the lookup region

> 7. `padded_byte` are copied to a word builder gadget to build padded words, which would later be copied to the `Keccak-f` permutation

Reviewer: This is a bit confusing. As some of the big points overlap the rows they touch. We could maybe group them.
Also, there's no mention to the special padding case with `0x81`

Author: For big points 1~5, those checks have to be applied on their specific offsets. 6 is another gadget and 7 is all copy constraints.
I just realized point 4 can be grouped. Let me check.
The `0x81` case is in the 5.1. If it is a pad zone and condition_80 is activated then it would be `0x81`.

Reviewer: Makes sense! Thanks!!!

Author: Addressed in [40e0186](https://github.com/appliedzkp/zkevm-specs/pull/147/commits/40e0186951315ab16eca26dfcb932f0b15b19933)

# Keccak256 For Variable Length Input

We describe the circuit that can process inputs of an arbitrary size.

## Plain Keccak behavior

Keccak256 uses a simple iterating sponge construction to handle a variable length input. The sponge absorbs 17 words(136 bytes) of inputs, applies the `Keccak-f` permutation, and then absorb next 17 words. Once all parts of the inputs are absorbed, the sponge squeezes out the 32 bytes for the output.

If the last part of the input is less than 136 bytes, a padding must be applied. The padding is a multi-rate padding, it pads single bit 1 followed by minimum number of bits 0 followed by a single bit 1, such that the result is 136 bytes.

## Circuit behavior

### Lookup Region

This region splits the input to multiple parts, each part corresponds to their `Keccak-f` permutation round.
This is also a lookup table for the other circuits to lookup the Keccak256 input to the output.

Define:

We use `Round` for the cells of the following columns in a row, it represents the information in a `keccak_f` round.

The `curr: Round` and `next: Round` represent the current and the next row.

Columns:

> - `hash_id` Sequential number for grouping different hash

Reviewer: I'm still not sure how can we have two `hash_id`s within the same circuit instance. In case we have it because all the permutations of all hashes will share some common region/table, couldn't we use the RLC(input) as id? I think makes a bit more sense as it ties the input to the permutation computations.

Author: Using input as the id might not make sense.
The definition of the input is "the RLC of the 136 bytes to be absorbed in this round". Let's say if we have two rows with the same input value 123,  their output could be different if one is the first 136 bytes of the acc input and the other is the second 136 bytes of the acc input.

Reviewer: But within the same permutation input computation, we already enforce the continuity and the relation between each byte by constraining the RLC.

If the RLC matches, the bytes were for sure from the same ID no?

- `input_len` Length for correct padding
- `input` 136 bytes to be absorbed in this round
- `acc_len` How many length we have absorbed

> - `acc_input` Accumulatd bytes by random linear combination (in big-endian order)

Reviewer: In this context, aren't these the same? We have absorbed as many bytes as we have accomulated in the RLC no?

Author: The `acc_len` is the length of bytes, while `acc_input` is RLC. `acc_len` can be 0 or a multiple of 136, we use it to track if a hash input ends at `input_len`.

Reviewer: ˋacc_inputˋ is always [0,135]??

- `output` The base-2 `state[:4]` output from this round `keccak_f`
- `output_word_[n]` 25 columns of the output in words of this round `keccak_f`
- `is_end_result` This flag indicates the hash output is final. The purpose is for external circuit to lookup.

Selector:

- `is_last_round_of_circuit` Last round of the whole circuit (not hash), to avoid `next` wrap around.

| hash_id | input_len | input | acc_len | acc_input | output | is_end_result | output_word_0 | output_word_1 | output_word_2 | output_word\_... |
| ------: | --------: | ----- | ------: | --------- | ------ | ------------: | ------------- | ------------- | ------------- | ---------------- |
|       0 |       150 |       |       0 |           |        |             0 |               |               |               |                  |

> |       0 |       150 |       |     136 |           |        |             1 |               |               |               |                  |

Reviewer: I'm not following this table at all. Is this related to an entire hash?? And here we have all the permutations of all of the hashes? If that's the case, Is this trying to describe the FULL keccak hash? Including N permutations per call?

Reviewer: If I'm correct, then `input` can't be a single column but rather 17 word columns or something similar no?

Author: Sorry, can you point to the columns of the cells again? I'm not following.
